### PR TITLE
Update env and sdk

### PR DESCRIPTION
**Pull request by sisuresh**
Draft PR for @leighmcculloch  to look at.

The liquidity pool test throws this error -
```
thread 'test::test' panicked at 'called `Result::unwrap()` on an `Err` value: HostError
Value: Status(HostFunctionError(InputArgsInvalid))

Debug events (newest first):
   0: "unexpected environment interface version"
```

### CODE REVIEW

**Comment by leighmcculloch** [comment]
Wow, something is breaking pretty hard. This is the generated code for the token contract in the liquidity contract. There's no functions there at all.

```rust
    pub trait Contract {}
    pub struct ContractClient {
        env: ::soroban_sdk::Env,
        contract_id: ::soroban_sdk::BytesN<32>,
    }
    impl ContractClient {
        pub fn new(
            env: &::soroban_sdk::Env,
            contract_id: impl ::soroban_sdk::IntoVal<::soroban_sdk::Env, ::soroban_sdk::BytesN<32>>,
        ) -> Self {
            Self {
                env: env.clone(),
                contract_id: contract_id.into_val(env),
            }
        }
    }
```

**Comment by leighmcculloch** [comment]
Indeed, it looks like no functions are exported:
```
❯ soroban-cli inspect --wasm soroban_token_contract.wasm
File: soroban_token_contract.wasm
Env Meta: AAAAAAAAAAAAAAAN
 • Interface Version: 13
Contract Spec: AAAAAQAAAAAAAAAQQWxsb3dhbmNlRGF0YUtleQAAAAIAAAAEZnJvbQAAB9AAAAAKSWRlbnRpZmllcgAAAAAAB3NwZW5kZXIAAAAH0AAAAApJZGVudGlmaWVyAAAAAAACAAAAAAAAAAdEYXRhS2V5AAAAAAgAAAAJQWxsb3dhbmNlAAAAAAAAAQAAB9AAAAAQQWxsb3dhbmNlRGF0YUtleQAAAAdCYWxhbmNlAAAAAAEAAAfQAAAACklkZW50aWZpZXIAAAAAAAVOb25jZQAAAAAAAAEAAAfQAAAACklkZW50aWZpZXIAAAAAAAVTdGF0ZQAAAAAAAAEAAAfQAAAACklkZW50aWZpZXIAAAAAAAVBZG1pbgAAAAAAAAAAAAAIRGVjaW1hbHMAAAAAAAAABE5hbWUAAAAAAAAABlN5bWJvbAAAAAAAAAAAAAEAAAAMc29yb2Jhbl9hdXRoAAAAEEVkMjU1MTlTaWduYXR1cmUAAAACAAAACnB1YmxpY19rZXkAAAAAA+4AAAAgAAAACXNpZ25hdHVyZQAAAAAAA+4AAABAAAAAAQAAAAxzb3JvYmFuX2F1dGgAAAARQWNjb3VudFNpZ25hdHVyZXMAAAAAAAACAAAACmFjY291bnRfaWQAAAAAA+4AAAAgAAAACnNpZ25hdHVyZXMAAAAAA+oAAAfQAAAAEEVkMjU1MTlTaWduYXR1cmUAAAACAAAADHNvcm9iYW5fYXV0aAAAAAlTaWduYXR1cmUAAAAAAAADAAAACENvbnRyYWN0AAAAAAAAAAdFZDI1NTE5AAAAAAEAAAfQAAAAEEVkMjU1MTlTaWduYXR1cmUAAAAHQWNjb3VudAAAAAABAAAH0AAAABFBY2NvdW50U2lnbmF0dXJlcwAAAAAAAAIAAAAMc29yb2Jhbl9hdXRoAAAACklkZW50aWZpZXIAAAAAAAMAAAAIQ29udHJhY3QAAAABAAAD7gAAACAAAAAHRWQyNTUxOQAAAAABAAAD7gAAACAAAAAHQWNjb3VudAAAAAABAAAD7gAAACAAAAABAAAADHNvcm9iYW5fYXV0aAAAABJTaWduYXR1cmVQYXlsb2FkVjAAAAAAAAQAAAAIZnVuY3Rpb24AAAAGAAAACGNvbnRyYWN0AAAD7gAAACAAAAAHbmV0d29yawAAAAAJAAAABGFyZ3MAAAPqAAAAAAAAAAIAAAAMc29yb2Jhbl9hdXRoAAAAEFNpZ25hdHVyZVBheWxvYWQAAAABAAAAAlYwAAAAAAABAAAH0AAAABJTaWduYXR1cmVQYXlsb2FkVjAAAA==
 • Struct: ScSpecUdtStructV0 { lib: VecM([]), name: VecM([65, 108, 108, 111, 119, 97, 110, 99, 101, 68, 97, 116, 97, 75, 101, 121]), fields: VecM([ScSpecUdtStructFieldV0 { name: VecM([102, 114, 111, 109]), type_: Udt(ScSpecTypeUdt { name: VecM([73, 100, 101, 110, 116, 105, 102, 105, 101, 114]) }) }, ScSpecUdtStructFieldV0 { name: VecM([115, 112, 101, 110, 100, 101, 114]), type_: Udt(ScSpecTypeUdt { name: VecM([73, 100, 101, 110, 116, 105, 102, 105, 101, 114]) }) }]) }
 • Union: ScSpecUdtUnionV0 { lib: VecM([]), name: VecM([68, 97, 116, 97, 75, 101, 121]), cases: VecM([ScSpecUdtUnionCaseV0 { name: VecM([65, 108, 108, 111, 119, 97, 110, 99, 101]), type_: Some(Udt(ScSpecTypeUdt { name: VecM([65, 108, 108, 111, 119, 97, 110, 99, 101, 68, 97, 116, 97, 75, 101, 121]) })) }, ScSpecUdtUnionCaseV0 { name: VecM([66, 97, 108, 97, 110, 99, 101]), type_: Some(Udt(ScSpecTypeUdt { name: VecM([73, 100, 101, 110, 116, 105, 102, 105, 101, 114]) })) }, ScSpecUdtUnionCaseV0 { name: VecM([78, 111, 110, 99, 101]), type_: Some(Udt(ScSpecTypeUdt { name: VecM([73, 100, 101, 110, 116, 105, 102, 105, 101, 114]) })) }, ScSpecUdtUnionCaseV0 { name: VecM([83, 116, 97, 116, 101]), type_: Some(Udt(ScSpecTypeUdt { name: VecM([73, 100, 101, 110, 116, 105, 102, 105, 101, 114]) })) }, ScSpecUdtUnionCaseV0 { name: VecM([65, 100, 109, 105, 110]), type_: None }, ScSpecUdtUnionCaseV0 { name: VecM([68, 101, 99, 105, 109, 97, 108, 115]), type_: None }, ScSpecUdtUnionCaseV0 { name: VecM([78, 97, 109, 101]), type_: None }, ScSpecUdtUnionCaseV0 { name: VecM([83, 121, 109, 98, 111, 108]), type_: None }]) }
 • Struct: ScSpecUdtStructV0 { lib: VecM([115, 111, 114, 111, 98, 97, 110, 95, 97, 117, 116, 104]), name: VecM([69, 100, 50, 53, 53, 49, 57, 83, 105, 103, 110, 97, 116, 117, 114, 101]), fields: VecM([ScSpecUdtStructFieldV0 { name: VecM([112, 117, 98, 108, 105, 99, 95, 107, 101, 121]), type_: BytesN(ScSpecTypeBytesN { n: 32 }) }, ScSpecUdtStructFieldV0 { name: VecM([115, 105, 103, 110, 97, 116, 117, 114, 101]), type_: BytesN(ScSpecTypeBytesN { n: 64 }) }]) }
 • Struct: ScSpecUdtStructV0 { lib: VecM([115, 111, 114, 111, 98, 97, 110, 95, 97, 117, 116, 104]), name: VecM([65, 99, 99, 111, 117, 110, 116, 83, 105, 103, 110, 97, 116, 117, 114, 101, 115]), fields: VecM([ScSpecUdtStructFieldV0 { name: VecM([97, 99, 99, 111, 117, 110, 116, 95, 105, 100]), type_: BytesN(ScSpecTypeBytesN { n: 32 }) }, ScSpecUdtStructFieldV0 { name: VecM([115, 105, 103, 110, 97, 116, 117, 114, 101, 115]), type_: Vec(ScSpecTypeVec { element_type: Udt(ScSpecTypeUdt { name: VecM([69, 100, 50, 53, 53, 49, 57, 83, 105, 103, 110, 97, 116, 117, 114, 101]) }) }) }]) }
 • Union: ScSpecUdtUnionV0 { lib: VecM([115, 111, 114, 111, 98, 97, 110, 95, 97, 117, 116, 104]), name: VecM([83, 105, 103, 110, 97, 116, 117, 114, 101]), cases: VecM([ScSpecUdtUnionCaseV0 { name: VecM([67, 111, 110, 116, 114, 97, 99, 116]), type_: None }, ScSpecUdtUnionCaseV0 { name: VecM([69, 100, 50, 53, 53, 49, 57]), type_: Some(Udt(ScSpecTypeUdt { name: VecM([69, 100, 50, 53, 53, 49, 57, 83, 105, 103, 110, 97, 116, 117, 114, 101]) })) }, ScSpecUdtUnionCaseV0 { name: VecM([65, 99, 99, 111, 117, 110, 116]), type_: Some(Udt(ScSpecTypeUdt { name: VecM([65, 99, 99, 111, 117, 110, 116, 83, 105, 103, 110, 97, 116, 117, 114, 101, 115]) })) }]) }
 • Union: ScSpecUdtUnionV0 { lib: VecM([115, 111, 114, 111, 98, 97, 110, 95, 97, 117, 116, 104]), name: VecM([73, 100, 101, 110, 116, 105, 102, 105, 101, 114]), cases: VecM([ScSpecUdtUnionCaseV0 { name: VecM([67, 111, 110, 116, 114, 97, 99, 116]), type_: Some(BytesN(ScSpecTypeBytesN { n: 32 })) }, ScSpecUdtUnionCaseV0 { name: VecM([69, 100, 50, 53, 53, 49, 57]), type_: Some(BytesN(ScSpecTypeBytesN { n: 32 })) }, ScSpecUdtUnionCaseV0 { name: VecM([65, 99, 99, 111, 117, 110, 116]), type_: Some(BytesN(ScSpecTypeBytesN { n: 32 })) }]) }
 • Struct: ScSpecUdtStructV0 { lib: VecM([115, 111, 114, 111, 98, 97, 110, 95, 97, 117, 116, 104]), name: VecM([83, 105, 103, 110, 97, 116, 117, 114, 101, 80, 97, 121, 108, 111, 97, 100, 86, 48]), fields: VecM([ScSpecUdtStructFieldV0 { name: VecM([102, 117, 110, 99, 116, 105, 111, 110]), type_: Symbol }, ScSpecUdtStructFieldV0 { name: VecM([99, 111, 110, 116, 114, 97, 99, 116]), type_: BytesN(ScSpecTypeBytesN { n: 32 }) }, ScSpecUdtStructFieldV0 { name: VecM([110, 101, 116, 119, 111, 114, 107]), type_: Bytes }, ScSpecUdtStructFieldV0 { name: VecM([97, 114, 103, 115]), type_: Vec(ScSpecTypeVec { element_type: Val }) }]) }
 • Union: ScSpecUdtUnionV0 { lib: VecM([115, 111, 114, 111, 98, 97, 110, 95, 97, 117, 116, 104]), name: VecM([83, 105, 103, 110, 97, 116, 117, 114, 101, 80, 97, 121, 108, 111, 97, 100]), cases: VecM([ScSpecUdtUnionCaseV0 { name: VecM([86, 48]), type_: Some(Udt(ScSpecTypeUdt { name: VecM([83, 105, 103, 110, 97, 116, 117, 114, 101, 80, 97, 121, 108, 111, 97, 100, 86, 48]) })) }]) }
```

**Comment by leighmcculloch** [comment]
There are two issues:

- You were right, the removal of the export feature in the token contract caused no functions to be exported. That's an easy fix.
- The second problem is feature unification is causing the `export` feature in on example to enable the `export` feature in the token contract... I didn't dig into why, I've just change the build script to use `cargo hack` so that each project is built in isolation.*

\* The reason I didn't dig into why is that I've lost days at this point to issues with cargo's feature unification and the treatment is usually the same: build the projects separately not together in workspace mode.

Fixes are in 391fd0c.

File: Makefile [hunk 1]
```diff
@@ -10,10 +10,16 @@ test: build-normal
 build: build-normal build-optimized
 
 build-normal:
-	cargo build --target wasm32-unknown-unknown --release
+	cargo build --target wasm32-unknown-unknown --release -p soroban-token-contract
+	cp target/wasm32-unknown-unknown/release/soroban_token_contract.wasm ./
+	cargo build --target wasm32-unknown-unknown --release -p soroban-liquidity-pool-contract
+	cp target/wasm32-unknown-unknown/release/soroban_liquidity_pool_contract.wasm ./
+	cargo build --target wasm32-unknown-unknown --release -p soroban-single-offer-contract
+	cp target/wasm32-unknown-unknown/release/soroban_single_offer_contract.wasm ./
```
Comment (leighmcculloch) [inline Makefile:18]: @sisuresh I added in `cp` commands to copy these files out on every build so we don't get into an issue where they are stale. I actually think you could change to just referencing the `target/` directories. I do that in some other examples and it works fine as long as the crate have a dependency in such a way they build in the right order, or they are named alphabetically to order that way. I didn't want to get into the weeds on that though, so I didn't make that change.

**Comment by sisuresh** [comment]
Thanks for fixing the issue @leighmcculloch.

**Comment by sisuresh** [comment]
Ah Looks like I can't approve my own PR.

**Review by leighmcculloch — APPROVED** [review]

**Pull request merged**